internal/cluster/client_message_test.go

```diff
@@ -0,0 +1,152 @@
+		{name: "readWhenTheFrameLengthAndFlagsNotReceivedAtFirst", f: readWhenTheFrameLengthAndFlagsNotReceivedAtFirst},
+		{name: "readFramesInMultipleCallsToReadFromWhenLastPieceIsSmall", f: readFramesInMultipleCallsToReadFromWhenLastPieceIsSmall},
+	}
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			tc.f(t)
+		})
+	}
+}
+
+func readSingleFrame(t *testing.T) {
+	frame := createFrameWithRandomBytes(t, 42)
+	message := proto.NewClientMessage(frame)
+	buffer := writeToBuffer(t, message)
+	reader := newClientMessageReader()
+	reader.Append(buffer.Bytes())
+	require.NotNil(t, reader.Read())
+	iterator := proto.NewForwardFrameIterator(reader.clientMessage.Frames)
+	require.True(t, iterator.HasNext())
+	frameRead := iterator.Next()
+	require.Equal(t, frame.Content, frameRead.Content)
+	require.False(t, iterator.HasNext())
+}
+
+func readMultiFrameMessage(t *testing.T) {
+	frame1 := createFrameWithRandomBytes(t, 10)
+	frame2 := createFrameWithRandomBytes(t, 20)
+	frame3 := createFrameWithRandomBytes(t, 30)
+	message := proto.NewClientMessageForEncode()
+	message.AddFrame(frame1)
+	message.AddFrame(frame2)
+	message.AddFrame(frame3)
+	buffer := writeToBuffer(t, message)
+	reader := newClientMessageReader()
+	reader.Append((*buffer).Bytes())
+	require.NotNil(t, reader.Read())
+	iterator := reader.clientMessage.FrameIterator()
+	require.True(t, iterator.HasNext())
+	frameRead := iterator.Next()
+	require.Equal(t, frame1.Content, frameRead.Content)
+	require.True(t, iterator.HasNext())
+	frameRead = iterator.Next()
+	require.Equal(t, frame2.Content, frameRead.Content)
+	require.True(t, iterator.HasNext())
+	frameRead = iterator.Next()
+	require.Equal(t, frame3.Content, frameRead.Content)
+	require.False(t, iterator.HasNext())
+}
+
+func readFramesInMultipleCallsToRead(t *testing.T) {
+	frame := createFrameWithRandomBytes(t, 1000)
+	message := proto.NewClientMessage(frame)
+	buffer := writeToBuffer(t, message)
+	// split message two part and send sequentially
+	part1 := buffer.Next(750)
+	part2 := buffer.Bytes()
+	part1buffer := bytes.NewBuffer(part1)
+	part2buffer := bytes.NewBuffer(part2)
+	reader := newClientMessageReader()
+	reader.Append(part1buffer.Bytes())
+	// should not finish reading
+	require.Nil(t, reader.Read())
+	reader.Append(part2buffer.Bytes())
+	// all the message has been read
+	require.NotNil(t, reader.Read())
+	iterator := reader.clientMessage.FrameIterator()
+	require.True(t, iterator.HasNext())
+	frameRead := iterator.Next()
+	require.Equal(t, frame.Content, frameRead.Content)
+	assert.False(t, iterator.HasNext())
+
+}
+
+func readWhenTheFrameLengthAndFlagsNotReceivedAtFirst(t *testing.T) {
+	frame := createFrameWithRandomBytes(t, 100)
+	message := proto.NewClientMessage(frame)
+	buffer := writeToBuffer(t, message)
+	reader := newClientMessageReader()
+	// should not be able to read with just 4 bytes of data
+	reader.Append((buffer.Bytes())[:4])
```

Comment:
I guess the parentheses around `buffer.Bytes()` is unnecessary. Can we remove it ?

Reply:
Removed at https://github.com/hazelcast/hazelcast-go-client/pull/863/commits/73051a7840e4f303b95e8fd0abc090403aff37b6.